qubes/tests/integ/vm_qrexec_gui.py

```diff
@@ -866,140 +819,152 @@ def test_300_bug_1028_gui_memory_pinning(self):
+        perror("mmap");
+        exit(1);
+    }
+
+    printf("Stage1\\n");
+    fflush(stdout);
+    getchar();
+    for (iter = addr; iter < addr + total_pages*0x1000; iter += 0x2000) {
+        if (mlock(iter, 0x1000) == -1) {
+            perror("mlock");
+            fprintf(stderr, "%d of %d\\n", (iter-addr)/0x1000, total_pages);
+            exit(1);
+        }
+    }
+
+    printf("Stage2\\n");
+    fflush(stdout);
+    for (iter = addr+0x1000; iter < addr + total_pages*0x1000; iter += 0x2000) {
+        if (munmap(iter, 0x1000) == -1) {
+            perror(\"munmap\");
+            exit(1);
+        }
+    }
+
+    printf("Stage3\\n");
+    fflush(stdout);
+    fclose(stdout);
+    getchar();
+
+    return 0;
+}
+'''
+
+        yield from self.testvm1.run_for_stdio('cat > allocator.c',
+            input=allocator_c.encode())
+
+        try:
+            stdout, stderr = yield from self.testvm1.run_for_stdio(
+                'gcc allocator.c -o allocator')
+        except subprocess.CalledProcessError:
+            self.skipTest('allocator compile failed: {}'.format(stderr))
 
         # drop caches to have even more memory pressure
-        self.testvm1.run("echo 3 > /proc/sys/vm/drop_caches",
-            user="root", wait=True)
+        yield from self.testvm1.run(
+            'echo 3 > /proc/sys/vm/drop_caches', user='root')
 
         # now fragment all free memory
-        p = self.testvm1.run("grep ^MemFree: /proc/meminfo|awk '{print $2}'",
-            passio_popen=True)
-        memory_pages = int(p.communicate()[0].strip())
-        memory_pages //= 4 # 4k pages
-        alloc1 = self.testvm1.run(
-            "ulimit -l unlimited; exec /home/user/allocator {}".format(
+        stdout, _ = yield from self.testvm1.run_for_stdio(
+            "grep ^MemFree: /proc/meminfo|awk '{print $2}'")
+        memory_pages = int(stdout) // 4  # 4k pages
+
+        alloc1 = yield from self.testvm1.run(
+            'ulimit -l unlimited; exec /home/user/allocator {}'.format(
                 memory_pages),
-            user="root", passio_popen=True, passio_stderr=True)
+            user="root")
+
         # wait for memory being allocated; can't use just .read(), because EOF
         # passing is unreliable while the process is still running
-        alloc1.stdin.write(b"\n")
-        alloc1.stdin.flush()
-        alloc_out = alloc1.stdout.read(len("Stage1\nStage2\nStage3\n"))
+        yield from alloc1.stdin.write(b'\n')
+        yield from alloc1.stdin.flush()
+        alloc_out = yield from alloc1.stdout.read(
+            len('Stage1\nStage2\nStage3\n'))
 
-        if b"Stage3" not in alloc_out:
+        if b'Stage3' not in alloc_out:
             # read stderr only in case of failed assert, but still have nice
             # failure message (don't use self.fail() directly)
-            self.assertIn(b"Stage3", alloc_out, alloc1.stderr.read())
+            #
+            # XXX why don't read stderr always? --woju 20170523
```

Comment:
Because on not-failed run, the process is still running, so stderr.read() will wait (indefinitely).